src/model.rs

```diff
@@ -159,14 +160,19 @@ impl Collections {
             }
         }
 
-        fn dedup_collection<T: PartialEq>(source: &mut Collection<T>) {
-            let mut dedup = Vec::default();
-            for object in source.take() {
-                if !dedup.contains(&object) {
-                    dedup.push(object);
-                }
+        fn dedup_collection<T: Clone + Eq + Hash>(source: &mut Collection<T>) -> Collection<T> {
+            let calculate_hash = |t: &T| -> u64 {
+                let mut s = DefaultHasher::new();
+                t.hash(&mut s);
+                s.finish()
+            };
+            let mut set: BTreeMap<u64, T> = BTreeMap::new();
+            let items = source.take();
+            for item in items {
+                set.insert(calculate_hash(&item), item);
             }
-            *source = Collection::new(dedup);
+            let collection: Vec<T> = set.values().cloned().collect();
+            Collection::new(collection)
```

Comment:
or we can use the crate https://docs.rs/itertools/0.10.1/itertools/index.html which uses a `HashMap(I::Item, ()>` where `I::Item` are the differents objects

```rust
use itertools::Itertools;
Collection::new(source.into_iter().unique().collect());
```